crates/sui-graphql-rpc/src/types/transaction_block/tx_lookups.rs

```diff
@@ -0,0 +1,468 @@
+    ) -> Self {
+        Self {
+            lo,
+            hi,
+            after,
+            before,
+            scan_limit,
+            is_from_front,
+        }
+    }
+
+    /// Determines the `tx_sequence_number` range from the checkpoint bounds for a transaction block
+    /// query. If no checkpoint range is specified, the default is between 0 and the
+    /// `checkpoint_viewed_at`. The corresponding `tx_sequence_number` range is fetched from db, and
+    /// further adjusted by cursors and scan limit. If the after cursor exceeds rhs, or before
+    /// cursor is below lhs, or other inconsistency, return None.
+    pub(crate) fn query(
+        conn: &mut Conn,
+        after_cp: Option<u64>,
+        at_cp: Option<u64>,
+        before_cp: Option<u64>,
+        checkpoint_viewed_at: u64,
+        scan_limit: Option<u64>,
+        page: &Page<Cursor>,
+    ) -> Result<Option<Self>, diesel::result::Error> {
+        let lo_cp = max_option!(after_cp.map(|x| x.saturating_add(1)), at_cp).unwrap_or(0);
+        let hi_cp = min_option!(
+            before_cp.map(|x| x.saturating_sub(1)),
+            at_cp,
+            Some(checkpoint_viewed_at)
+        )
+        .unwrap();
+        let from_db: StoredTxBounds =
+            conn.result(move || tx_bounds_query(lo_cp, hi_cp).into_boxed())?;
+
+        let lo = from_db.lo as u64;
+        let hi = from_db.hi as u64;
+
+        if page.after().map_or(false, |x| x.tx_sequence_number >= hi)
+            || page.before().map_or(false, |x| x.tx_sequence_number <= lo)
+        {
+            return Ok(None);
+        }
+
+        Ok(Some(Self::new(
+            lo,
+            hi,
+            page.after().map(|x| x.tx_sequence_number),
+            page.before().map(|x| x.tx_sequence_number),
+            scan_limit,
+            page.is_from_front(),
+        )))
+    }
+
+    /// The lower bound `tx_sequence_number`` of the range to scan within. This defaults to the min
+    /// tx_sequence_number of the checkpoint bound. If a cursor is provided, the lower bound is
+    /// adjusted to the larger of the two. The resulting value is additionally modified by the
+    /// `scan_limit` if `is_from_front` is false.
+    pub(crate) fn scan_lo(&self) -> u64 {
+        let adjusted_lo = self.after.map_or(self.lo, |a| std::cmp::max(self.lo, a));
+
+        if self.is_from_front {
+            adjusted_lo
+        } else {
+            std::cmp::max(
+                adjusted_lo,
+                self.scan_hi()
+                    .saturating_sub(self.scan_limit.unwrap_or(self.hi)),
+            )
+        }
+    }
+
+    /// The upper bound `tx_sequence_number` of the range to scan within. This defaults to the max
+    /// tx_sequence_number of the checkpoint bound. If a cursor is provided, the upper bound is
+    /// adjusted to the smaller of the two. The resulting value is additionally modified by the
+    /// `scan_limit` if `is_from_front` is true.
+    pub(crate) fn scan_hi(&self) -> u64 {
```

Comment:
The co-recursion here is making me nervous -- let's split this function into two more parts -- `tx_lo` and `tx_hi` that give the bounds before the scan limit, and then `scan_lo` and `scan_hi` that give the limits after adjusting for the scan limit.

Reply:
Also other comments from `scan_lo` apply equally here.

Reply:
Sounds good to me. Logically, if scan_lo depends on scan_hi, then scan_hi is a simple return, and vice versa, but it's not very legible today and quite confusing